### ISSUE TITLE
Writing JS strings to resources, Deno.encode and Deno.encodeInto

### ISSUE THREAD

**Issue by billywhizz** [denoland/deno#15895]
## Issue

I was doing some investigation into network perf using Deno FFI and stdio perf in general and came up against a performance roadblock which is relevant to anywhere we are writing JS strings to ops resources using send or write syscalls and the like.

If you look at the benchmark script [here](https://github.com/denoland/performance-team/blob/main/benchmarks/io/02-strings/bench.js#L34) you can see the following scenarios benchmarked. All scenarios write to /dev/null using FFI fast call.

## Benchmarks

- **static_buffer**: writing a pre-existing buffer with the string pre-encoded - nothing should be faster than this afaik
- **encode_dynamic_string**: writing a dynamically generated string encoded into a new Uint8Array using ```Deno.core.encode```
- **encode_static_string**: writing a static string encoded into a new Uint8Array using ```Deno.core.encode```
- **encode_into_dynamic_string**: writing a dynamically generated string encoded into an existing Uint8Array using ```Deno.core.ops.op_encoding_encode_into```
- **encode_into_static_string**: writing a static string encoded into an existing Uint8Array using ```Deno.core.ops.op_encoding_encode_into```

## Results
![results](https://user-images.githubusercontent.com/159073/189983780-471f7b07-883c-47db-a1fd-8c8fac50e66f.png)

## Flamegraphs

Flamegraphs to compare the three static string scenarios are here which show clearly the overhead involved in both ```Deno.core.encode``` and ```Deno.core.ops.op_encoding_encode_into``` compared to the baseline of writing a pre-encoded buffer.

### static_buffer
![static_buffer](https://user-images.githubusercontent.com/159073/189983512-be9f2082-612b-4a14-b81b-233bd3230840.svg)

### encode_static_string
![encode_static_string](https://user-images.githubusercontent.com/159073/189983520-390f9a41-b1e2-4453-9417-59d20004734d.svg)

### encode_into_static_string
![encode_into_static_string](https://user-images.githubusercontent.com/159073/189983526-fdf079ca-e0e9-4cae-aad2-e014518f986b.svg)

## Discussion

From the benchmarks we can see:
- calling ```Deno.core.encode``` on the static string in question causes a **~58%** drop in throughput and takes **2.3x** longer
- calling ```Deno.core.ops.op_encoding_encode_into``` causes a **~64%** drop in throughput and takes **3.27x** longer
- JS String concatentation adds **~15%** overhead

From the flamegraphs we can see:
- with a static buffer, the program spends **88%** of time in ```__libc_write```
- with ```Deno.core.encode``` the program spends **35%** in ```__libc_write```
- with ```Deno.core.ops.op_encoding_encode_into``` the program spends **28%** in ```__libc_write```

There are a few of issues I think this raises:
- ```Deno.core.encode``` is expensive because it does a lot of work - it allocated a BackingStore, creates an ArrayBuffer and wraps it in a Uint8Array and returns that to JS. This is per spec so not much we can do about it other than to avoid using it where we don't need to.
-  ```Deno.core.ops.op_encoding_encode_into``` **should** be faster than Deno.core.encode as it is working with a pre-existing buffer and only needs to do a memcpy from the string into the buffer. The overhead here is likely down to it being a native op and being wrapped in serde_v8.
- We need a method for writing a JS string directly to an ops resource rid. I came across the same issue with just-js and had to do something like this. This is significantly quicker than first encoding into a buffer and then writing the buffer as far as i can remember.

```C++
  String::Utf8Value str(isolate, args[1]);
  args.GetReturnValue().Set(Integer::New(isolate, write(fd, 
    *str, str.length())));
```

## Planned Actions

- add an [op-v8] api call ```Deno.core.encode_into (u8, String)``` into ops_builtin_v8.rs which does encode_into without serde_v8.
- add a ```Deno.core.write_utf8 (rid, string)``` api call to [ops_builtin_v8.rs](https://github.com/denoland/deno/blob/main/core/ops_builtin_v8.rs) which will take a copy of the utf8 bytes inside the v8 string passed to it and write them to the resource identified by rid, as in the C++ snippet above.

The disadvantage of the second approach is, afaik, it will not be optimised as a fast call because fast call api has no string support yet. But it would be worth knowing if it is faster than the other methods above even without the fast call so I propose doing it as a draft PR and see if it improves perf. It should be pretty easy to implement.

**Comment by aapoalas** [comment]
Don't you worry: Fast API can actually work with strings. If you give the parameter type as `kValue` then it'll accept any type (presumably) and gets passed as the V8 wrapper Value.

**Comment by billywhizz** [comment]
nice. has that landed already @aapoalas? i'll give it a go if so.

**Comment by aapoalas** [comment]
Yeah, it's already supported by Fast API internally. I can't quite remember if the support for that in ops was already added.

You'll need to use the options bag parameter as well to fall back to the slow path when the expected string parameter turns out to be not-a-string.

**Comment by littledivy** [comment]
For the record, here's the link to the conversation on V8 supporting strings in Fast API calls. 
https://discord.com/channels/684898665143206084/778060818046386176/1008754326933618728

**Comment by billywhizz** [comment]
i've updated the OP with a new test using Deno.writeSync, which is fastest "official" way I can find to write to a resource. Also uploaded nicer flamegraphs with full Deno/Rust stacktraces.

**Comment by littledivy** [comment]
After applying the encodeInto optimization in #15922 

![encodeInto_ptr](https://user-images.githubusercontent.com/34997667/190560487-c687baff-5e6f-405e-96ed-4bd70d596eb5.svg)

**Comment by littledivy** [comment]
The `Builtins_IC` is concerning. I think its related to webidl?

**Comment by billywhizz** [comment]
I've implemented a little [minimal runtime](https://github.com/billywhizz/deno-runtime-demo) so i can test these low level ops changes without building all of deno. it only takes ~20 seconds to do a rebuild with this.

I implemented @littledivy's change from [this PR](https://github.com/denoland/deno/pull/15922/) for encode_into as well as a similar change with the same logic current Deno uses.

Divy's PR is ~3x the current implementation in throughput on a small http response payload. I will benchmark again over weekend using this in the benchmarks i did for text writing above.
![Screenshot from 2022-09-17 03-04-51](https://user-images.githubusercontent.com/159073/190836375-8019c5f0-c131-48ed-8b7b-2d46dd3aeb3e.png)

Here are the flamegraphs

### current encode_into
![deno_encode_into](https://user-images.githubusercontent.com/159073/190836387-c8556ef6-8ec7-4617-ad0b-33760630a80e.svg)

### my [change](https://github.com/billywhizz/deno-runtime-demo/blob/main/src/main.rs#L51) based on divy's PR
![runjs_encode_into](https://user-images.githubusercontent.com/159073/190836402-e9517e84-7936-4565-9222-174cb84f06ba.svg)

### divy's PR [change](https://github.com/billywhizz/deno-runtime-demo/blob/main/src/main.rs#L29)
![runjs_encode_into_fast](https://user-images.githubusercontent.com/159073/190836413-e72f3ad8-1476-472b-8bfb-dae002d9e928.svg)

**Comment by billywhizz** [comment]
btw - i just noticed the fast implementation is still using GetBackingStore. i'll have a look at that tomorrow. I might not be picking up latest rusty_v8.
![Screenshot from 2022-09-17 03-22-08](https://user-images.githubusercontent.com/159073/190836843-cad665f6-b950-4e9f-9652-e2f9c8f9c22a.png)

**Comment by littledivy** [comment]
@billywhizz you need to use `deno_ops` and `deno_core` from my PR branch. The `GetBackingStore` changes are unreleased.

**Comment by littledivy** [comment]
great results though!

**Comment by bartlomieju** [comment]
This issue has been largely addressed over the past few years. Here's a summary of the current state (Deno 2.x, measured on Apple Silicon):

## What changed

- `op_encoding_encode_into` now has a proper fast call path (`#[op2(fast)]`) bypassing serde_v8 overhead — this was the main fix from #15922
- `Deno.core.encode` is no longer exposed in Deno 2.x; users use `TextEncoder` directly
- `TextEncoder.encodeInto` is now very fast, using direct V8 string APIs

## Current benchmarks (5M iterations, ~83 byte HTTP response string)

| Operation | Deno (ops/s) | Node (ops/s) | Deno vs Node |
|---|---|---|---|
| `TextEncoder.encode` | **4.62M** | 3.64M | **1.27x faster** |
| `TextEncoder.encodeInto` | **22.47M** | 11.53M | **1.95x faster** |
| `Buffer.from(string)` | 7.71M | 13.67M | 0.56x (slower) |
| `writeSync(preEncoded)` | 2.41M | 1.82M | 1.32x faster |
| `encodeInto + writeSync` | 2.08M | — | — |
| `writeSync(string)` (node compat) | 1.71M | 1.52M | 1.13x faster |

The main bottleneck identified in the original report (`encodeInto` being slow due to serde_v8 wrapping) has been resolved. `TextEncoder.encodeInto` is now ~2x faster than Node.

The proposed `write_utf8(rid, string)` API was never implemented, but the practical impact is minimal given how fast `encodeInto` is now — the encoding step adds only ~14% overhead on top of `writeSync` with a pre-encoded buffer.

Closing as addressed.
